README.md
<div align="center">
<img src="https://raw.githubusercontent.com/lantunes/netomaton/master/resources/logo.png" alt="logo"></img>
</div>

<b>Netomaton</b> is a Python framework for exploring discrete dynamical systems. It is a software abstraction
meant to aid in the implementation of models of collective computation. Examples of such computational models include 
Cellular Automata and Neural Networks. This also includes some continuous dynamical systems, such as ordinary and 
partial differential equations, since the simulation of such systems involves the discretization of space and time. 
Netomaton is also a tool for exploring Complex Systems.

Underlying all discrete dynamical systems (and discretized continuous dynamical systems) are networks of stateful units 
that obey rules that specify how their states change over time. Netomaton thus considers all dynamical systems as a 
model of computation known as Functional Network Automata.

[![testing status](https://github.com/lantunes/netomaton/actions/workflows/python-app.yml/badge.svg?branch=master)](https://github.com/lantunes/netomaton/actions)
[![latest version](https://img.shields.io/pypi/v/netomaton?style=flat-square&logo=PyPi&logoColor=white&color=blue)](https://pypi.org/project/netomaton/)

### Getting Started

Netomaton can be installed via pip:

```
pip install netomaton
```

Requirements for using this library are Python 3.6, numpy 1.15.4,
matplotlib 3.0.2, networkx 2.5, and msgpack 1.0.2.


### What are Network Automata?

The [Wikipedia entry](https://en.wikipedia.org/wiki/Network_automaton)
for Network Automata has stated:

> A network automaton (plural network automata) is a mathematical system consisting of a network of nodes that evolves over time according to predetermined rules. It is similar in concept to a cellular automaton, but much less studied.

> Stephen Wolfram's book _A New Kind of Science_, which is primarily concerned with cellular automata, briefly discusses network automata, and suggests (without positive evidence) that the universe might at the very lowest level be a network automaton.

A Network Automaton is a discrete dynamical system comprised of a collection
of nodes (the computational units) causally connected to each other, as
specified by a network-defining adjacency matrix. The nodes adopt states
at each timestep of the network's evolution, as prescribed by an activity
function, *f*. Moreover, the network's topology can also change over time, as
prescribed by a connectivity function, *g*.

The network's topology is specified by the adjacency matrix, **A**, which
is of size _N_<sub>tot</sub> *X* _N_<sub>tot</sub>, where _N_<sub>tot</sub>
represents the total number of nodes in the network. Each non-zero entry
in **A** represents the existence of a link. The value of the entry
represents a link weight. The matrix **A** thus contains information
about the existence of links, and their direction.

The network is evolved for *T* timesteps. The activity of the network is
defined by the activities of all its nodes, and is represented by **S**<sub>*t*</sub>,
where *t* is a particular timestep. During each timestep, the activity
function *f* is invoked, followed by the connectivity function *g*, such
that:

**S**<sub>*t+1*</sub> = *f*(**A**<sub>*t*</sub>, **S**<sub>*t*</sub>)

**A**<sub>*t+1*</sub> = *g*(**A**<sub>*t*</sub>, **S**<sub>*t*</sub>)


There are no restrictions to the kinds of topological changes that a
network may undergo over the course of its evolution. A network may have
nodes added or removed at any given timestep.

To learn more, please refer to the scientific literature on the subject:

> Wolfram, S. (2002). A New Kind of Science (pp. 475–545). Champaign, IL: Wolfram Media.

> Tomassini, Marco. "Generalized automata networks." International Conference on Cellular Automata. Springer, Berlin, Heidelberg, 2006.

> Sayama, Hiroki, and Craig Laramee. "Generative network automata: A generalized framework for modeling adaptive network dynamics using graph rewritings." Adaptive Networks. Springer, Berlin, Heidelberg, 2009. 311-332.

> Smith, David MD, et al. "Network automata: Coupling structure and function in dynamic networks." Advances in Complex Systems 14.03 (2011): 317-339.

### Examples

Here's an example of the Elementary Cellular Automaton Rule 30 (as
described by Stephen Wolfram in his book
[_A New Kind of Science_](https://www.wolframscience.com/nks/)),
implemented with the Netomaton library:
```python
import netomaton as ntm

network = ntm.topology.cellular_automaton(n=200)

initial_conditions = [0] * 100 + [1] + [0] * 99

trajectory = ntm.evolve(network=network, initial_conditions=initial_conditions,
                        activity_rule=ntm.rules.nks_ca_rule(30), timesteps=100,
                        memoize=True)

ntm.plot_activities(trajectory)
```

<img src="resources/rule30.png" width="50%"/>

This repository contains examples of implementations of
various kinds of collective computation models, all implemented using
the Netomaton framework. Follow the link to learn more:

* [Elementary Cellular Automata](demos/elementary_ca/README.md)
* [1D Cellular Automata with Totalistic Rules](demos/totalistic_ca/README.md)
* [Reversible 1D Cellular Automata](demos/reversible_ca/README.md)
* [Density Classification with Evolved 1D Cellular Automata](demos/ca_density_classification/README.md)
* [Density Classification with a Watts-Strogatz small-world graph](demos/small_world_density_classification/README.md)
* [Asynchronous Automata](demos/asynchronous_automata/README.md)
* [Continuous Automata](demos/continuous_automata/README.md)
* [Finite State Machines](demos/finite_state_machine/README.md)
* [Pushdown Automata](demos/pushdown_automata/README.md)
* [Turing Machines](demos/turing_machine/README.md)
* [Langton's Lambda and Measures of Complexity](demos/langtons_lambda/README.md)
* [2D Cellular Automata](demos/totalistic_ca2d/README.md)
* [Conway's Game of Life](demos/game_of_life/README.md)
* [Fredkin's Self-Replicating CA](demos/fredkin_self_replicating_ca/README.md)
* [Langton's Loops](demos/langtons_loops/README.md)
* [Gray-Scott Reaction-Diffusion Model](demos/reaction_diffusion/README.md)
* [Hexagonal Cell Lattices](demos/hexagonal_ca/README.md)
* [Hopfield Network](demos/hopfield_net/README.md)
* [Restricted Boltzmann Machine](demos/restricted_boltzmann_machine/README.md)
* [Multilayer Perceptron](demos/mlp/README.md)
* [Perturbations](demos/perturbation_demo/README.md)
* [Sandpiles](demos/sandpiles/README.md)
* [Continuous-Time Models](demos/continuous_time_models/README.md)
* [Travelling Salesman Problem with the Hopfield-Tank Neural Net](demos/hopfield_tank_tsp/README.md)
* [Logistic Map](demos/logistic_map/README.md)
* [Lorenz Attractor](demos/lorenz_attractor/README.md)
* [Collatz Conjecture](demos/collatz_conjecture/README.md)
* [Substitution Systems](demos/substitution_systems/README.md)
* [Lindenmayer Systems](demos/lindenmayer_systems/README.md)
* [Wireworld](demos/wireworld/README.md)
* [Random Attachment Model](demos/random_attachment_model/README.md)
* [Randomly Growing Network](demos/randomly_growing_network/README.md)
* [Restricted Network Automata](demos/restricted_network_automata/README.md)
* [Functional Network Automata](demos/functional_network_automata/README.md)
* [Evolving Networks](demos/evolving_networks/README.md)
* [Fungal Growth Model](demos/fungal_growth/README.md)
* [Flocking](demos/flocking/README.md)
* [Optimizing Particle Swarms](demos/optimizing_particle_swarms/README.md)
* [Wolfram Physics Model](demos/wolfram_physics/README.md)

Additionally, this library includes a number of utility functions for
working with the results produced by the automata. For example, there
is the `animate` function, which is explained more [here](demos/animation_demo/README.md).
It is also important to understand the `timesteps` and `input`
parameters of the `evolve` function, explained [here](demos/timesteps_and_input/README.md).

### About this project

This project proposes the idea that many popular and well-known
collective computational models can all be thought of as Functional Network Automata.
Such models include Cellular Automata, Boltzmann Machines, and various
flavours of Neural Networks, such as the Hopfield Net, and the Multilayer
Perceptron. This library does not attempt to be a replacement for great
frameworks such as TensorFlow, which are optimized, both in software and
hardware, for working with Neural Networks, for example. What this library
does attempt to be is a generalization of collective computation,
instantiated in software, with the goal of helping us see similarities
between models, and imagine new models that borrow features from existing
examples. It aims to provide a software architecture for understanding
and exploring the nature of computation in potentially dynamic networks.

Netomaton arose from a personal need to reconcile various models of collective
computation. In what fundamental ways does a Neural Network differ from a
Cellular Automaton? What can a Boltzmann Machine do that other models can't?
What do any of these models have in common? What sorts of new models can
we imagine? These are the questions that this library aspires to help answer.

Netomaton tries to make accessible any model of collective computation.
In so doing, it adopts certain generalizations and abstractions that,
while providing a common language for discussing seemingly disparate kinds of
models, incur a cost in terms of increased runtime complexity. The cost
of being very general is a less than ideal runtime performance, as any
given implementation is not optimized for a specific setting. For
example, regarding neural networks roughly as a series of matrix
multiplications allows one to take advantage of software and hardware
that can do those operations quickly. The focus of Netomaton, on the
other hand, is not on practicality, but on flexibility.

### Development

Create a Conda environment from the provided environment YAML file:
```
$ conda env create -f netomaton_dev.yaml
```

**Documentation**

To build the Sphinx documentation, from the `doc` directory:
```
$ make clean html
```
The generated files will be in `_build/html`.

**Testing**

There are a number of unit tests for this project. To run the tests:
```
$ pytest tests
```

--------------------

### Citation Info

This project has been published on [Zenodo](https://zenodo.org/record/3893141#.XuUOg55KhZI),
which provides a DOI, as well as an easy way to generate citations in a number of formats.
For example, this project may be cited as:


> Antunes, Luis M. (2019, September 28). Netomaton: A Python Library for working with 
Network Automata. Zenodo. http://doi.org/10.5281/zenodo.3893141


BibTeX:
```
@software{antunes_luis_m_2019_3893141,
  author       = {Antunes, Luis M.},
  title        = {{Netomaton: A Python Library for working with 
                   Network Automata}},
  month        = sep,
  year         = 2019,
  publisher    = {Zenodo},
  doi          = {10.5281/zenodo.3893141},
  url          = {https://doi.org/10.5281/zenodo.3893141}
}
```

### Stars

Please star this repository if you find it useful, or use it as part of your research.

### Copyrights

Copyright (c) 2018-2020 Luis M. Antunes (@lantunes) All rights reserved.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]


## [1.3.0] - 2023-01-14

### Added

- Added `bipartite` function to topology module
- Added `complete` function to topology module
- Added `lattice_pos` function to vis module
- Added demo for flocking
- Added demo for restricted Boltzmann machine
- Added demo for optimizing particle swarms
- Added demo for multilayer perceptron
- Added demo for Lorenz attractor

### Changed

- Changed `animate_network`: added `with_timesteps` and `show` parameters
- Changed `animate_activities`: added `with_timesteps` and `blit` parameters

## [1.2.0] - 2022-12-22

### Added

- Added `CHANGELOG.md`
- Added `MemoizationKey` and memoization support
- Added demo for functional network automata
- Added demo for evolving networks

### Changed

- Moved version to `__netomaton/__init__.py`
- Explicitly importing symbols instead of using `*` in `__ini__.py` files
- Making `_bits_to_int` and `_int_to_bits` public
- Changed `binary_ca_rule`: no longer shifting neighbourhood activities to center
- Changed `Network`: adding `rotation_system` property
- Changed `evolution.py`: adding support for memoization and rotation systems

### Removed

- Removed `shift_to_center`

## [1.1.2] - 2022-05-25

### Changed

- Fixed problem in `game_of_life_rule`

## [1.1.1] - 2021-08-01

### Changed

- Loosened some dependency requirements in `setup.py`

## [1.1.0] - 2021-07-24

### Added

- Added support for CTRBL rules
- Added Langton's Loop implementation

## [1.0.0] - 2021-07-04

### Added

- Initial stable release


netomaton/utils.py
import networkx as nx
import math
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation
import matplotlib.collections as mcoll
import collections
from .state import State
from .substitution_system import SubstitutionSystem
try:
    import cPickle as pickle
except:
    import pickle


def plot_activities(trajectory_or_activities, shape=None, slice=-1, title='', colormap='Greys', vmin=None, vmax=None,
                    node_annotations=None, show_grid=False):
    if len(trajectory_or_activities) is 0:
        raise Exception("there are no activities")
    if isinstance(trajectory_or_activities[0], State):
        # trajectory_or_activities is a trajectory, convert it to an activities list
        activities = get_activities_over_time_as_list(trajectory_or_activities)
    else:
        activities = trajectory_or_activities
    if shape is not None:
        activities = _reshape_for_animation(activities, shape)
    plot_grid(activities, shape, slice, title, colormap, vmin, vmax, node_annotations, show_grid)


def get_activities_over_time_as_list(trajectory):
    activities = []
    for state in trajectory:
        ac = state.activities
        row = []
        if ac:
            for a in sorted(ac):
                row.append(ac[a])
            activities.append(row)
    return activities


def plot_grid(activities, shape=None, slice=-1, title='', colormap='Greys', vmin=None, vmax=None,
              node_annotations=None, show_grid=False):
    if shape is not None:
        activities = np.array(activities).reshape((len(activities), shape[0], shape[1]))[slice]
    cmap = plt.get_cmap(colormap)
    plt.title(title)
    plt.imshow(activities, interpolation='none', cmap=cmap, vmin=vmin, vmax=vmax)

    if node_annotations is not None:
        for i in range(len(node_annotations)):
            for j in range(len(node_annotations[i])):
                plt.text(j, i, node_annotations[i][j], ha="center", va="center", color="grey",
                         fontdict={'weight':'bold','size':6})

    if show_grid:
        plt.grid(which='major', axis='both', linestyle='-', color='grey', linewidth=0.5)
        plt.xticks(np.arange(-.5, len(activities[0]), 1), "")
        plt.yticks(np.arange(-.5, len(activities), 1), "")
        plt.tick_params(axis='both', which='both', length=0)

    plt.show()


def plot_grid_multiple(ca_list, shape=None, slice=-1, titles=None, colormap='Greys', vmin=None, vmax=None):
    cmap = plt.get_cmap(colormap)
    for i in range(0, len(ca_list)):
        plt.figure(i)
        if titles is not None:
            plt.title(titles[i])
        activities = list(ca_list[i])
        if shape is not None:
            activities = np.array(activities).reshape((len(activities), shape[0], shape[1]))[slice]
        plt.imshow(activities, interpolation='none', cmap=cmap, vmin=vmin, vmax=vmax)
    plt.show()


def animate_activities(trajectory_or_activities, title='', shape=None, save=False, interval=50, colormap='Greys',
                       vmin=None, vmax=None, show_grid=False, show_margin=True, scale=0.6, dpi=80, blit=True,
                       with_timestep=False):
    if len(trajectory_or_activities) is 0:
        raise Exception("there are no activities")
    if isinstance(trajectory_or_activities[0], State):
        # trajectory_or_activities is a trajectory, convert it to an activities list
        activities = get_activities_over_time_as_list(trajectory_or_activities)
    else:
        activities = trajectory_or_activities
    if shape is not None:
        activities = _reshape_for_animation(activities, shape)
    cmap = plt.get_cmap(colormap)
    fig, ax = plt.subplots()
    title_text = plt.title(title)
    if not show_margin:
        fig.subplots_adjust(left=0, bottom=0, right=1, top=1, wspace=0, hspace=0)

    grid_linewidth = 0.0
    if show_grid:
        plt.xticks(np.arange(-.5, len(activities[0][0]), 1), "")
        plt.yticks(np.arange(-.5, len(activities[0]), 1), "")
        plt.tick_params(axis='both', which='both', length=0)
        grid_linewidth = 0.5

    vertical = np.arange(-.5, len(activities[0][0]), 1)
    horizontal = np.arange(-.5, len(activities[0]), 1)
    lines = ([[(x, y) for y in (-.5, horizontal[-1])] for x in vertical] +
             [[(x, y) for x in (-.5, vertical[-1])] for y in horizontal])
    grid = mcoll.LineCollection(lines, linestyles='-', linewidths=grid_linewidth, color='grey')
    ax.add_collection(grid)

    im = plt.imshow(activities[0], animated=True, cmap=cmap, vmin=vmin, vmax=vmax)
    if not show_margin:
        baseheight, basewidth = im.get_size()
        fig.set_size_inches(basewidth*scale, baseheight*scale, forward=True)

    i = {'index': 0}
    def updatefig(*args):
        i['index'] += 1
        if i['index'] == len(activities):
            i['index'] = 0
        im.set_array(activities[i['index']])
        if with_timestep:
            title_text.set_text("timestep: %s" % (i['index']+1))
        return im, grid, title_text
    ani = animation.FuncAnimation(fig, updatefig, interval=interval, blit=blit, save_count=len(activities))
    if save:
        ani.save('evolved.gif', dpi=dpi, writer="imagemagick")
    plt.show()


def animate_plot1D(x, y, save=False, interval=50, dpi=80):
    if isinstance(y[0], State):
        # y is a trajectory, convert it to an activities list
        y = get_activities_over_time_as_list(y)
    fig1 = plt.figure()
    line, = plt.plot(x, y[0])
    def update_line(activity):
        line.set_data(x, activity)
        return line,
    ani = animation.FuncAnimation(fig1, update_line, frames=y, blit=True, interval=interval)
    if save:
        ani.save('plot.gif', dpi=dpi, writer="imagemagick")
    plt.show()


def _reshape_for_animation(activities, shape):
    if len(shape) == 1:
        assert shape[0] == len(activities[0]), "shape must equal the length of an activity vector"
        new_activities = []
        for i, a in enumerate(activities):
            new_activity = []
            new_activity.extend(activities[0:i+1])
            while len(new_activity) < len(activities):
                new_activity.append([0]*len(activities[0]))
            new_activities.append(new_activity)
        return np.array(new_activities)
    elif len(shape) == 2:
        return np.reshape(activities, (len(activities), shape[0], shape[1]))
    else:
        raise Exception("shape must be a tuple of length 1 or 2")


def plot_network(network, layout="shell", with_labels=True, node_color="#1f78b4", node_size=300):
    G = network.to_networkx()
    if layout == "shell":
        nx.draw_shell(G, with_labels=with_labels, node_color=node_color, node_size=node_size)
    elif layout == "spring":
        nx.draw_spring(G, with_labels=with_labels, node_color=node_color, node_size=node_size)
    elif isinstance(layout, dict):
        nx.draw(G, pos=layout, with_labels=with_labels, node_color=node_color, node_size=node_size)
    else:
        raise Exception("unsupported layout: %s" % layout)
    plt.show()


def animate_network(trajectory, save=False, interval=50, dpi=80, layout="shell",
                    with_labels=True, with_arrows=True, node_color="b", node_size=30, with_timestep=False, show=True):
    fig, ax = plt.subplots()

    def update(arg):
        ax.clear()

        i, state = arg

        if with_timestep:
            ax.set_title("timestep: %s" % (i+1))

        color = node_color[i] if type(node_color) == dict else node_color

        network = state.network
        G = nx.MultiDiGraph()
        for n in network.nodes:
            G.add_node(n)
        for edge in network.edges:
            G.add_edge(edge[0], edge[1])

        state_layout = layout[i] if type(layout) == list or type(layout) == tuple else layout

        if state_layout == "shell":
            nx.draw_shell(G, with_labels=with_labels, node_color=color, node_size=node_size, arrows=with_arrows)
        elif state_layout == "spring":
            nx.draw_spring(G, with_labels=with_labels, node_color=color, node_size=node_size, arrows=with_arrows)
        elif state_layout == "planar":
            nx.draw_planar(G, with_labels=with_labels, node_color=color, node_size=node_size, arrows=with_arrows)
        elif state_layout == "kamada-kawai":
            nx.draw_kamada_kawai(G, with_labels=with_labels, node_color=color, node_size=node_size, arrows=with_arrows)
        elif isinstance(state_layout, dict):
            nx.draw(G, pos=state_layout, with_labels=with_labels, node_color=color, node_size=node_size, arrows=with_arrows)
        else:
            raise Exception("unsupported layout: %s" % state_layout)

    ani = animation.FuncAnimation(fig, update, frames=list(enumerate(trajectory)), interval=interval,
                                  save_count=len(trajectory))
    if save:
        ani.save('evolved.gif', dpi=dpi, writer="imagemagick")
    if show:
        plt.show()

    return ani


def plot1D(x, y, color=None, label=None, xlabel=None, ylabel=None, xlim=None, ylim=None, twinx=False,
           legend=None, tight_layout=None, title=None):
    """
    Creates a 1D plot of the given x and y values.

    :param x: A list representing the values of the x-axis.

    :param y: A list representing the values of the y-axis. If the values of this list are also lists, then each will
              be plotted as a separate series on the y-axis.

    :param color: A string, representing the color of the series, or a list of colors, representing the color of each
                   series to be plotted. The number of colors must match the number of series. The color values must
                   be recognizable by Matplotlib.

    :param label: A string or number, or a list of strings or numbers, representing the labels of each series to be
                  plotted. The number of labels must match the number of series.

    :param xlabel: A string representing the label of the x-axis.

    :param ylabel: A string, or list of at most two strings, representing the label(s) of the y-axis.

    :param xlim: A 2-tuple of numbers, representing the limits of the x-axis.

    :param ylim: A 2-tuple of numbers, or a list of at most two 2-tuples of numbers, representing the
                 limits of the y-axis.

    :param twinx: If True, the provided y series will be plotted on two separate y-axes, and there must
                  be at least two series of y values provided. If there are more than two series of y values,
                  then each series will be plotted on alternating y-axes. (Default is False)

    :param legend: A dict, or a list of at most two dicts, representing the arguments to the legend
                   function of Matplotlib.

    :param tight_layout: A dict representing the arguments to the tight_layout function of Matplotlib.

    :param title: The plot title.
    """
    axes = []
    fig, ax1 = plt.subplots()
    axes.append(ax1)
    if twinx:
        axes.append(ax1.twinx())
        if len(y) < 2:
            raise Exception("there must be at least two series of y values provided")
        for _y in y:
            if not isinstance(_y, (list, np.ndarray)):
                raise Exception("an item in y must be a list representing a y series")

    current_axis_idx = 0
    # assume that the first value type in y is representative of all the value types in y
    is_multiseries = isinstance(y[0], (list, np.ndarray))

    def _get_item(items, i):
        if isinstance(items, str):
            return items
        if isinstance(items, collections.Sequence):
            return items[i]
        return items

    if not is_multiseries:
        y = [y]

    for i, y_series in enumerate(y):
        current_axis = axes[current_axis_idx % len(axes)]
        plot_args = {}
        if label:
            plot_args["label"] = _get_item(label, i)
        if color:
            plot_args["color"] = _get_item(color, i)
        current_axis.plot(x, y_series, **plot_args)
        if xlabel:
            current_axis.set_xlabel(xlabel)
        if ylabel:
            ylabel_args = {}
            if color:
                ylabel_args["color"] = _get_item(color, i)
            current_axis.set_ylabel(_get_item(ylabel, i % 2), **ylabel_args)
        if xlim:
            current_axis.set_xlim(xlim)
        if ylim:
            current_axis.set_ylim(_get_item(ylim, i % 2))
        if legend:
            current_axis.legend(**legend)

        current_axis_idx += 1

    if tight_layout:
        plt.tight_layout(**tight_layout)

    if title:
        plt.title(title)

    plt.show()


def poincare_plot(activities, timesteps, xlabel=None, ylabel=None, xlim=None, ylim=None, title=None):
    """
    Create a Poincaré plot.

    :param activities: A list of activities. If the values of this list are also lists, then each will
                       be plotted as a separate series.

    :param timesteps: The number of timesteps in the trajectory to consider, starting from the end.

    :param xlabel: A string representing the label of the x-axis.

    :param ylabel: A string representing the label of the y-axis.

    :param xlim: A 2-tuple of numbers, representing the limits of the x-axis.

    :param ylim: A 2-tuple of numbers, or a list of at most two 2-tuples of numbers, representing the
                 limits of the y-axis.

    :param title: The plot title.
    """
    cm = plt.get_cmap('gist_rainbow')
    fig = plt.figure()
    ax = fig.add_subplot(111)
    # assume that the first value type in y is representative of all the value types in y
    is_multiseries = isinstance(activities[0], (list, np.ndarray))
    if is_multiseries:
        ax.set_prop_cycle(color=[cm(1. * i / len(activities)) for i in range(len(activities))])
    else:
        activities = [activities]
    for a in activities:
        x = []
        y = []
        for t in range(timesteps - 1):
            x.append(a[-timesteps:][t])
            y.append(a[-timesteps:][t + 1])
        plt.scatter(x, y, s=1)
    if xlim:
        plt.xlim(xlim)
    if ylim:
        plt.ylim(ylim)
    if xlabel:
        plt.xlabel(xlabel)
    if ylabel:
        plt.ylabel(ylabel)
    if title:
        plt.title(title)
    plt.show()


def bifurcation_plot(x, trajectories, timesteps, xlabel=None, ylabel=None, title=None):
    """
    Create a bifurcation plot.

    :param x: A list representing the values of the x-axis.

    :param trajectories: A list of lists. Each inner list is a sequence of activities representing a trajectory.

    :param timesteps: The number of timesteps in the trajectory to consider, starting from the end.

    :param xlabel: A string representing the label of the x-axis.

    :param ylabel: A string representing the label of the y-axis.

    :param title: The plot title.
    """
    y = []
    for t in trajectories:
        y.append(np.unique(t[-timesteps:]))
    for x_e, y_e in zip(x, y):
        plt.scatter([x_e] * len(y_e), y_e, color="b", s=1)
    if xlabel:
        plt.xlabel(xlabel)
    if ylabel:
        plt.ylabel(ylabel)
    if title:
        plt.title(title)
    plt.show()


def plot_degree_distribution(network, xlabel="Node degree", ylabel_freq="Frequency", ylabel_prob="Probability",
                             in_degree=False, out_degree=False,
                             equation=None, equation_x=0.51, equation_y=0.76, equation_text="", equation_color="r",
                             color="r", title=None):
    """
    Create a node degree distribution plot for the given network.

    :param network: A Netomaton Network instance.

    :param xlabel: The x-axis label.

    :param ylabel_freq: The frequency y-axis label.

    :param ylabel_prob: The probability y-axis label.

    :param in_degree: If True, the in-degree will be used. (default is False)

    :param out_degree: If True, the out-degree will be used. (default is False)

    :param equation: A callable that computes the degree distribution, given a node degree.

    :param equation_x: The equation's x coordinate.

    :param equation_y: The equation's y coordinate.

    :param equation_text: The equation to display.

    :param equation_color: The equation text's color. It must be a valid Matplotlib color.

    :param color: The color to use for the plot. It must be a valid Matplotlib color.

    :param title: The plot's title.
    """
    degree_counts = {}
    for node in network.nodes:
        if in_degree:
            degree = network.in_degree(node)
        elif out_degree:
            degree = network.out_degree(node)
        else:
            degree = network.degree(node)

        if degree not in degree_counts:
            degree_counts[degree] = 0
        degree_counts[degree] += 1

    x = [i for i in range(1, max(degree_counts) + 1)]
    height = [degree_counts[i] if i in degree_counts else 0 for i in x]
    plt.bar(x, height)
    plt.xlabel(xlabel)
    plt.ylabel(ylabel_freq)

    if equation:
        y = [equation(k) for k in x]
        plt.twinx()
        plt.plot(x, y, color=color)
        plt.ylabel(ylabel_prob)
        plt.text(equation_x, equation_y, equation_text, transform=plt.gca().transAxes, color=equation_color)

    if title:
        plt.title(title)

    plt.show()


def ncr(n, r):
    """
    Returns the number of ways to choose r items from n items without repetition and without order.

    :param n: A non-negative integer.

    :param r: A non-negative integer.

    :return: An integer representing the number of ways to choose r items from n items without repetition
             and without order.
    """
    if not isinstance(n, int) or n < 0:
        raise Exception("n must be a non-negative int")
    if not isinstance(r, int) or r < 0:
        raise Exception("r must be a non-negative int")
    if n < r:
        return 0
    f = math.factorial
    return f(n) // f(r) // f(n-r)


def plot_L_system(state, turtle, bindings, **kwargs):
    _render_L_system(state, bindings)
    turtle.plot(**kwargs)


def animate_L_system(state, turtle, bindings, **kwargs):
    _render_L_system(state, bindings)
    turtle.animate(**kwargs)


def _render_L_system(state, bindings):
    state = SubstitutionSystem.to_string([state])[0]
    for s in state:
        instructions = bindings[s]
        if not isinstance(instructions, list):
            instructions = [instructions]
        for instruction in instructions:
            if callable(instruction):
                instruction()
            else:
                fn, *args = instruction
                fn(*args)


def binarize_for_plotting(activities):
    """
    Expects a list of lists of single numbers representing the activities (e.g. [[2], [35], [12]]),
    and returns a list of lists of the binarized versions of the given numbers, left-padded with
    zeroes. (e.g. [[0, 0, 0, 0, 1, 0], [1, 0, 0, 0, 1, 1], [0, 0, 1, 1, 0, 0]])

    :param activities: the activities as a list of lists of single numbers

    :return: a list of lists of the binary versions of the numbers, left-padded with zeroes
    """
    activities = [[int(x) for x in bin(int(a[0]))[2:]] for a in activities]
    max_len = np.max([len(a) for a in activities])
    return np.asarray([np.pad(a, (max_len - len(a), 0), 'constant', constant_values=0) for a in activities])
